app/app.go

```diff
@@ -188,7 +189,11 @@ func (a *App) startLeaderElection(ctx context.Context, configMapsGetter core_v1c
 		cancel()
 		return nil, nil, err
 	}
-	go le.Run()
+	go func() {
+		// note: because le.Run() also adds a logging panic handler panics with be logged 3 times
```

Comment:
Hm, we can actually use that [`HandleCrash()`](https://github.com/kubernetes/apimachinery/blob/release-1.11/pkg/util/runtime/runtime.go#L48) function instead of deferring our own.

While at it, change [`ErrorHandlers `](https://github.com/kubernetes/apimachinery/blob/release-1.11/pkg/util/runtime/runtime.go#L86) to use our logger? I don't think it is used anywhere in client libraries we use (although I haven't checked) but it may be later.

Reply:
I saw that nasty bit of mutable global state and am hesitant to randomly reach out an change it. (it also looks like they might remove it in future). Is the suggestion to write `runtime.panicHandlers = []func(interface{}){logStructuredPanic}` and 

```
runtime.ErrorHandlers = []func(error){
	logStructuredPanic,
	(&rudimentaryErrorBackoff{
		lastErrorTime: time.Now(),
		// 1ms was the number folks were able to stomach as a global rate limit.
		// If you need to log errors more than 1000 times a second you
		// should probably consider fixing your code instead. :)
		minPeriod: time.Millisecond,
	}).OnError,
}
```
That second func is hard to work around without copying the code or pretending like slice[1] is API. If you're suggesting something like `runtime.panicHandlers = append(panicHandler, logStructuredPanic)` that's less crazy although still gross and brittle across upgrades. 

Are all the current logging cases covered by changing this global state?

Reply:
> I saw that nasty bit of mutable global state and am hesitant to randomly reach out and change it

It is nasty, but it is ok to change it if you do it from the main goroutine (e.g. from `init()`) before any other ones are started. Yes, the example you provided is what I'm suggesting.

> Are all the current logging cases covered by changing this global state?

Most likely no, but quite a lot of code in our vendor dir calls `HandleError(err)`. However, I doubt we are hitting this handler a lot if at all. So I wouldn't bother with `rudimentaryErrorBackoff()`.

Reply:
> I'd prefer to manage it by making ctrl (and our internal code) wrap the go routines they create.

Yes, but that means we'll have things logged multiple times, in a structured and in an unstructured format. This is probably more confusing than just leaving it alone. But then we are not solving the problem you want to solve (migrate to structured logging).